runtime/lua/vim/lsp/buf.lua

```diff
@@ -152,11 +151,14 @@ end
 --
 ---@see https://microsoft.github.io/language-server-protocol/specification#textDocument_formatting
 function M.formatting(options)
-  local client = select_client("textDocument/formatting")
-  if client == nil then return end
+  select_client('textDocument/formatting', function(client)
+    if client == nil then
+      return
+    end
 
-  local params = util.make_formatting_params(options)
-  return client.request("textDocument/formatting", params, nil, vim.api.nvim_get_current_buf())
+    local params = util.make_formatting_params(options)
+    return client.request('textDocument/formatting', params, nil, vim.api.nvim_get_current_buf())
```

Comment:
I think both the `util.make_formatting_params(options)` and the `vim.api.nvim_get_current_buf()` might need to move outside of the `on_choice` callback to ensure it uses the document buffer and not the buffer or window used for the `vim.ui.select` input.

It depends a bit on whether the `vim.ui.select` implementation closes its buffer/window first before calling `on_choice`. Just to be on the safe side.

Same for the other methods.

Reply:
I don't think where on_choice is run is specified anywhere, so I implemented the recommended change